tests/integration/cocoon/test_utilities.py

```diff
@@ -28,3 +31,111 @@ def test_get_swagger_dict_fail(self, _, api_url, expected_exception):
 
         with self.assertRaises(expected_exception):
             cocoon.utilities.get_swagger_dict(api_url=api_url)
+
```

Comment:
There is a missing test case for:

- Providing a string of a for a model which doesn't exist e.g. "NotALusidModel"
- Providing an empty list
- Providing an index greater than the length of the list

Reply:
Moved test from integration to unit.

Reply:
(2) Added all three recommended tests

Reply:
Great nice work! For Providing an empty list you should check that the error message matches your expectation as well. In this case "The provided list of attribute_for_grouping is empty".

For:
- Providing a string of a for a model which doesn't exist e.g. "NotALusidModel"
- Providing an index greater than the length of the list

We need to make sure that we are providing an error which helps the user (even if an internal function) understand what has gone wrong. These two tests only assert that the default unhandled error is raised which is unlikely to help a user understand where the problem can be fixed. You should add some better handling for these cases in your utilities.group_request_into_one function so that you control the error that is raised.

Reply:
You will need to check that the error message is as you expect. Your check for the index being greater than the length of the list actually comes after you try and use that index so the IndexError that you get is not the one you wanted to send with a helpful error.

Reply:
Good example here - https://github.com/finbourne/lusid-sdk-python-preview/blob/master/sdk/tests/utilities/test_api_client_factory.py#L48-L50